## 19.4.2 DeepLearning_Tabular

In [1]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler,OneHotEncoder
import pandas as pd
import tensorflow as tf

# Import our input dataset
attrition_df = pd.read_csv('Resources/HR-Employee-Attrition.csv')
attrition_df.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   ...  RelationshipSatisfaction StandardHours  StockOptionLevel  \
0  ...                         1            80                 0   
1  ...                         4            80                 1   
2  ...                         2            80                 0   
3  ...                         3            80                 0   
4  ...                         4            80                 1   

   TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                  8                      0               1               6   
1                 10                      3               3              10   
2                  7                      3               3               0   
3                  8                      3               3               8   
4                  6                      3               3               2   

  YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
0                  4                        0                     5  
1                  7                        1                     7  
2                  0                        0                     0  
3                  7                        3                     0  
4                  2                        2                     2  

[5 rows x 35 columns]

In [2]:
# Generate our categorical variable list
attrition_cat = attrition_df.dtypes[attrition_df.dtypes == "object"].index.tolist()

## 19.4.3 Real Data, Real Practice Preprocessing

In [3]:
# Check the number of unique values in each column.
attrition_df[attrition_cat].nunique()

Attrition         2
BusinessTravel    3
Department        3
EducationField    6
Gender            2
JobRole           9
MaritalStatus     3
Over18            1
OverTime          2
dtype: int64

In [4]:
# Create a OneHotEncoder instance
enc = OneHotEncoder(sparse=False)

# Fit and transform the OneHotEncoder using the categorical variable list
encode_df = pd.DataFrame(enc.fit_transform(attrition_df[attrition_cat]))

# Add the encoded variable names to the DataFrame
encode_df.columns = enc.get_feature_names(attrition_cat)
encode_df.head()

Attrition_No  Attrition_Yes  BusinessTravel_Non-Travel  \
0           0.0            1.0                        0.0   
1           1.0            0.0                        0.0   
2           0.0            1.0                        0.0   
3           1.0            0.0                        0.0   
4           1.0            0.0                        0.0   

   BusinessTravel_Travel_Frequently  BusinessTravel_Travel_Rarely  \
0                               0.0                           1.0   
1                               1.0                           0.0   
2                               0.0                           1.0   
3                               1.0                           0.0   
4                               0.0                           1.0   

   Department_Human Resources  Department_Research & Development  \
0                         0.0                                0.0   
1                         0.0                                1.0   
2                         0.0                                1.0   
3                         0.0                                1.0   
4                         0.0                                1.0   

   Department_Sales  EducationField_Human Resources  \
0               1.0                             0.0   
1               0.0                             0.0   
2               0.0                             0.0   
3               0.0                             0.0   
4               0.0                             0.0   

   EducationField_Life Sciences  ...  JobRole_Research Director  \
0                           1.0  ...                        0.0   
1                           1.0  ...                        0.0   
2                           0.0  ...                        0.0   
3                           1.0  ...                        0.0   
4                           0.0  ...                        0.0   

   JobRole_Research Scientist  JobRole_Sales Executive  \
0                         0.0                      1.0   
1                         1.0                      0.0   
2                         0.0                      0.0   
3                         1.0                      0.0   
4                         0.0                      0.0   

   JobRole_Sales Representative  MaritalStatus_Divorced  \
0                           0.0                     0.0   
1                           0.0                     0.0   
2                           0.0                     0.0   
3                           0.0                     0.0   
4                           0.0                     0.0   

   MaritalStatus_Married  MaritalStatus_Single  Over18_Y  OverTime_No  \
0                    0.0                   1.0       1.0          0.0   
1                    1.0                   0.0       1.0          1.0   
2                    0.0                   1.0       1.0          0.0   
3                    1.0                   0.0       1.0          0.0   
4                    1.0                   0.0       1.0          1.0   

   OverTime_Yes  
0           1.0  
1           0.0  
2           1.0  
3           1.0  
4           0.0  

[5 rows x 31 columns]

In [5]:
# Merge one-hot encoded features and drop the originals
attrition_df = attrition_df.merge(encode_df,left_index=True, right_index=True)
attrition_df = attrition_df.drop(attrition_cat,1)
attrition_df.head()

Age  DailyRate  DistanceFromHome  Education  EmployeeCount  EmployeeNumber  \
0   41       1102                 1          2              1               1   
1   49        279                 8          1              1               2   
2   37       1373                 2          2              1               4   
3   33       1392                 3          4              1               5   
4   27        591                 2          1              1               7   

   EnvironmentSatisfaction  HourlyRate  JobInvolvement  JobLevel  ...  \
0                        2          94               3         2  ...   
1                        3          61               2         2  ...   
2                        4          92               2         1  ...   
3                        4          56               3         1  ...   
4                        1          40               3         1  ...   

   JobRole_Research Director  JobRole_Research Scientist  \
0                        0.0                         0.0   
1                        0.0                         1.0   
2                        0.0                         0.0   
3                        0.0                         1.0   
4                        0.0                         0.0   

   JobRole_Sales Executive  JobRole_Sales Representative  \
0                      1.0                           0.0   
1                      0.0                           0.0   
2                      0.0                           0.0   
3                      0.0                           0.0   
4                      0.0                           0.0   

   MaritalStatus_Divorced  MaritalStatus_Married  MaritalStatus_Single  \
0                     0.0                    0.0                   1.0   
1                     0.0                    1.0                   0.0   
2                     0.0                    0.0                   1.0   
3                     0.0                    1.0                   0.0   
4                     0.0                    1.0                   0.0   

   Over18_Y  OverTime_No  OverTime_Yes  
0       1.0          0.0           1.0  
1       1.0          1.0           0.0  
2       1.0          0.0           1.0  
3       1.0          0.0           1.0  
4       1.0          1.0           0.0  

[5 rows x 57 columns]

In [6]:
# Split our preprocessed data into our features and target arrays
y = attrition_df["Attrition_Yes"].values
X = attrition_df.drop(["Attrition_Yes","Attrition_No"],1).values

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=78)

In [7]:
# Create a StandardScaler instance
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## 19.4.4 Real Data, Real Practice Deep Learning Model Design

In [8]:
# Define the model - deep neural net
number_input_features = len(X_train[0])
hidden_nodes_layer1 = 8
hidden_nodes_layer2 = 5

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(
    tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation="relu")
)

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation="relu"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 8)                 448       
                                                                 
 dense_1 (Dense)             (None, 5)                 45        
                                                                 
 dense_2 (Dense)             (None, 1)                 6         
                                                                 
Total params: 499
Trainable params: 499
Non-trainable params: 0
_________________________________________________________________


In [9]:
# Compile the model
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

## 19.4.5 Real Data, Real Practice Train and Evaluate the Model

In [10]:
# Train the model
fit_model = nn.fit(X_train,y_train,epochs=100)

Epoch 1/100
35/35 [==============================] - 0s 770us/step - loss: 385.6817 - accuracy: 0.8122
Epoch 2/100
35/35 [==============================] - 0s 690us/step - loss: 152.9973 - accuracy: 0.6878
Epoch 3/100
35/35 [==============================] - 0s 687us/step - loss: 84.7207 - accuracy: 0.6779
Epoch 4/100
35/35 [==============================] - 0s 751us/step - loss: 69.2695 - accuracy: 0.7232
Epoch 5/100
35/35 [==============================] - 0s 721us/step - loss: 56.6224 - accuracy: 0.7187
Epoch 6/100
35/35 [==============================] - 0s 768us/step - loss: 47.3846 - accuracy: 0.7214
Epoch 7/100
35/35 [==============================] - 0s 754us/step - loss: 38.2071 - accuracy: 0.7178
Epoch 8/100
35/35 [==============================] - 0s 804us/step - loss: 31.9716 - accuracy: 0.7441
Epoch 9/100
35/35 [==============================] - 0s 744us/step - loss: 23.9151 - accuracy: 0.7577
Epoch 10/100
35/35 [==============================] - 0s 943us/step - loss: 18.0

35/35 [==============================] - 0s 742us/step - loss: 1.5255 - accuracy: 0.7368
Epoch 82/100
35/35 [==============================] - 0s 726us/step - loss: 0.9524 - accuracy: 0.7849
Epoch 83/100
35/35 [==============================] - 0s 772us/step - loss: 0.9824 - accuracy: 0.8094
Epoch 84/100
35/35 [==============================] - 0s 787us/step - loss: 1.2316 - accuracy: 0.7677
Epoch 85/100
35/35 [==============================] - 0s 747us/step - loss: 2.0699 - accuracy: 0.7514
Epoch 86/100
35/35 [==============================] - 0s 745us/step - loss: 1.8677 - accuracy: 0.7514
Epoch 87/100
35/35 [==============================] - 0s 738us/step - loss: 1.5163 - accuracy: 0.7659
Epoch 88/100
35/35 [==============================] - 0s 724us/step - loss: 0.7154 - accuracy: 0.8258
Epoch 89/100
35/35 [==============================] - 0s 752us/step - loss: 1.0728 - accuracy: 0.7922
Epoch 90/100
35/35 [==============================] - 0s 731us/step - loss: 1.3638 - accuracy: 

In [11]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

12/12 - 0s - loss: 3.1003 - accuracy: 0.8668 - 92ms/epoch - 8ms/step
Loss: 3.1003377437591553, Accuracy: 0.866847813129425


In [12]:
# Define the model - deep neural net
number_input_features = len(X_train_scaled[0])
hidden_nodes_layer1 = 10
hidden_nodes_layer2 = 5

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(
    tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation="relu")
)

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation="relu"))


# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Compile the Sequential model together and customize metrics
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [13]:
# Train the model 
fit_model = nn.fit(X_train_scaled, y_train, epochs=50) 
# Evaluate the model using the test data 
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

Epoch 1/50
35/35 [==============================] - 0s 765us/step - loss: 0.6804 - accuracy: 0.5209
Epoch 2/50
35/35 [==============================] - 0s 797us/step - loss: 0.5978 - accuracy: 0.7051
Epoch 3/50
35/35 [==============================] - 0s 820us/step - loss: 0.5373 - accuracy: 0.7922
Epoch 4/50
35/35 [==============================] - 0s 812us/step - loss: 0.4887 - accuracy: 0.8194
Epoch 5/50
35/35 [==============================] - 0s 833us/step - loss: 0.4524 - accuracy: 0.8240
Epoch 6/50
35/35 [==============================] - 0s 806us/step - loss: 0.4257 - accuracy: 0.8285
Epoch 7/50
35/35 [==============================] - 0s 850us/step - loss: 0.4058 - accuracy: 0.8303
Epoch 8/50
35/35 [==============================] - 0s 786us/step - loss: 0.3905 - accuracy: 0.8312
Epoch 9/50
35/35 [==============================] - 0s 769us/step - loss: 0.3775 - accuracy: 0.8312
Epoch 10/50
35/35 [==============================] - 0s 791us/step - loss: 0.3673 - accuracy: 0.8321

### SVM and deep learning models both achieved a predictive accuracy of around 87%.